imagepuller/internal/service/service.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2025 Edgeless Systems GmbH
+// SPDX-License-Identifier: BUSL-1.1
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/storage"
+	"github.com/containers/storage/types"
+	"github.com/edgelesssys/contrast/imagepuller/internal/api"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"golang.org/x/sys/unix"
+)
+
+// ImagePullerService is the struct for which the PullImage ttRPC service is implemented.
+type ImagePullerService struct {
+	Logger *slog.Logger
+}
+
+// PullImage is a ttRPC service which pulls and mounts docker images.
+func (s *ImagePullerService) PullImage(ctx context.Context, r *api.ImagePullRequest) (response *api.ImagePullResponse, retErr error) {
+	log := s.Logger.With(
+		slog.String("image_url", r.ImageUrl),
+		slog.String("bundle_path", r.BundlePath),
+	)
+	log.Info("Handling image pull request")
+
+	defer func() {
+		if retErr != nil {
+			log.Error("request failed", "err", retErr)
+			response = nil
+		}
+	}()
+
+	ref, err := name.ParseReference(r.ImageUrl)
+	if err != nil {
+		return nil, fmt.Errorf("parsing the image URL as a references: %w", err)
+	}
+
+	tr := transport.NewRetry(remote.DefaultTransport)
+	remoteImg, err := remote.Image(ref, remote.WithContext(ctx), remote.WithTransport(tr))
+	if errors.Is(err, context.Canceled) || errors.Is(ctx.Err(), context.Canceled) {
```

Comment:
I think that the second condition is not needed:

1. If the context expired during the call, the returned error is likely to wrap that context error.
2. If it only expired afterwards, we could just as well continue until the next remote call.

You could also consider `context.Timeout` as a special error case here.

Reply:
> `context.Timeout`

`context.DeadlineExceeded`?